File: occasum-core/src/main/java/pw/stamina/occasum/scan/field/FieldIgnorePredicate.java
package pw.stamina.occasum.scan.field;

import pw.stamina.occasum.scan.field.model.Ignore;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.util.Objects;
import java.util.function.Predicate;

public interface FieldIgnorePredicate extends Predicate<Field> {

    boolean shouldIgnoreField(Field field);

    @Override
    default FieldIgnorePredicate or(Predicate<? super Field> other) {
        Objects.requireNonNull(other);
        return (t) -> test(t) || other.test(t);
    }

    @Override
    default boolean test(Field field) {
        return shouldIgnoreField(field);
    }

    static FieldIgnorePredicate standard() {
        return ifIgnored().or(ifStatic()).or(ifNotFinal());
    }

    static FieldIgnorePredicate ifStatic() {
        return field -> Modifier.isStatic(field.getModifiers());
    }

    static FieldIgnorePredicate ifNotFinal() {
        return field -> !Modifier.isFinal(field.getModifiers());
    }

    static FieldIgnorePredicate ifIgnored() {
        return field -> field.isAnnotationPresent(Ignore.class);
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/scan/field/FieldIgnorePredicate.kt
package pw.stamina.occasum.scan.field

import pw.stamina.occasum.scan.field.model.Ignore
import java.lang.reflect.Field
import java.lang.reflect.Modifier

typealias FieldIgnorePredicate = (Field) -> Boolean

object FieldIgnorePredicates {

    infix fun FieldIgnorePredicate.or(other: FieldIgnorePredicate): FieldIgnorePredicate {
        return { field -> this(field) || other(field) }
    }

    fun standard(): FieldIgnorePredicate {
        return ifIgnored() or ifStatic() or ifNotFinal()
    }

    fun ifStatic(): FieldIgnorePredicate {
        return { field -> Modifier.isStatic(field.modifiers) }
    }

    fun ifNotFinal(): FieldIgnorePredicate {
        return { field -> !Modifier.isFinal(field.modifiers) }
    }

    fun ifIgnored(): FieldIgnorePredicate {
        return { field -> field.isAnnotationPresent(Ignore::class.java) }
    }
}
